### PR TITLE
Task/2fa with email

### PR DESCRIPTION
**Pull request by AlvaroVega**
Details: https://github.com/telefonicaid/fiware-keystone-spassword/blob/task/2fa_with_email/docs/second_factor_auth.md
- [x] Implement new APIs
- [x] Tests for APIs
- [x] Check LDAP users: should be excluded

### CODE REVIEW

File: keystone-spassword.spec [hunk 1]
```diff
@@ -63,12 +63,16 @@ pwd_max_tries=5
 pwd_block_minutes=30
```
Comment (fgalan) [inline keystone-spassword.spec]: CNR entry missing 

(I know the PR is WIP and may pass some time until it gets ready to merge... but better comment this sooner than latter so I don't forget  about it ;)

Comment (AlvaroVega) [reply keystone-spassword.spec]: This PR is far to be ready for merge

File: docs/second_factor_auth.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Second Factor Authentication (2FA)
+
+According with [wikipedia](https://en.wikipedia.org/wiki/Multi-factor_authentication) Two-factor authentication (also known as 2FA) is a type (subset) of multi-factor authentication. It is a method of confirming a user's claimed identity by utilizing a combination of two different factors: 1) something they know, 2) something they have, or 3) something they are.
+
+This feature provider 2FA for OpenStack Keystone based on email.
```
Comment (fgalan) [inline docs/second_factor_auth.md]: Typo: provider -> provides

File: docs/second_factor_auth.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+Specific options at /etc/keystone/keystone.conf
+```
+[spassword]
+sndfa=true
+sndfa_time_window=24
+sndfa_endpoint='localhost:5001'
+```
+
+`sndfa` is a boolean which enables (true) or disables (false) if Second Factor Authentication feature is available in Keystone instance.
```
Comment (fgalan) [inline docs/second_factor_auth.md]: Formatting suggestion: use a bullet list for this.

File: docs/second_factor_auth.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+[spassword]
+smtp_server=localhost
+smtp_port=587
+smtp_tls=true
+smtp_user='smtpuser@yourdomain.com'
+smtp_password='yourpassword'
+smtp_from='smtpuser'
+```
+
+## API
```
Comment (fgalan) [inline docs/second_factor_auth.md]: It seems there are some problems rendering this section:

imagen

Not sure if  `<this>` is correctly rendered in github... you can use code-style (i.e. using ```) or use `{alternative}`.

File: keystone-spassword.spec (hunk 1 above)
Comment (fgalan) [reply keystone-spassword.spec]: NTC (it was added in the last times)

File: keystone_spassword/contrib/spassword/core.py [hunk 5]
```diff
@@ -159,32 +190,32 @@ def update_user_modification_time(self, user):
...
+# class SPasswordMiddleware(wsgi.Middleware):
 
-    def __init__(self, *args, **kwargs):
-        LOG.debug("SPasswordMiddleware INIT")
-        try:
-            self.spassword_api = SPasswordManager()
-        except Exception:
-            LOG.debug("SPasswordMiddleware already registered")
-        return super(SPasswordMiddleware, self).__init__(*args, **kwargs)
+#     def __init__(self, *args, **kwargs):
```
Comment (fgalan) [inline keystone_spassword/contrib/spassword/core.py]: Leftover comment block?

File: keystone_spassword/contrib/spassword/mailer.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+#
+# Copyright 2015 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline keystone_spassword/contrib/spassword/mailer.py]: New file? Living in the past (2015) ? :)

File: docs/second_factor_auth.md (hunk 2 above)
Comment (AlvaroVega) [reply docs/second_factor_auth.md]: Fixed in: https://github.com/telefonicaid/fiware-keystone-spassword/pull/57/commits/d97ef84d0e7f433fe8eac697c526156bf89648b1

File: keystone_spassword/contrib/spassword/core.py [hunk 7]
```diff
@@ -25,8 +25,7 @@
 from keystone import auth
 from keystone.auth.plugins import password
 from keystone.common import dependency
-from keystone.common import extension
-from keystone.common import wsgi
+#from keystone.common import wsgi
```
Comment (fgalan) [inline keystone_spassword/contrib/spassword/core.py]: Leftover comment?

Comment (AlvaroVega) [reply keystone_spassword/contrib/spassword/core.py]: Fixed in https://github.com/telefonicaid/fiware-keystone-spassword/pull/57/commits/09846cb1bead252b70340d4a8f033c750aa0cd3f

File: keystone_spassword/contrib/spassword/mailer.py (hunk 6 above)
Comment (AlvaroVega) [reply keystone_spassword/contrib/spassword/mailer.py]: Fixed in: https://github.com/telefonicaid/fiware-keystone-spassword/pull/57/commits/76dd25a8f329e9dc480df3834a6e239c86ca4f4e

File: keystone_spassword/contrib/spassword/core.py (hunk 5 above)
Comment (AlvaroVega) [reply keystone_spassword/contrib/spassword/core.py]: Fixed in: https://github.com/telefonicaid/fiware-keystone-spassword/pull/57/commits/09846cb1bead252b70340d4a8f033c750aa0cd3f

File: docs/second_factor_auth.md (hunk 4 above)
Comment (AlvaroVega) [reply docs/second_factor_auth.md]: fixed in https://github.com/telefonicaid/fiware-keystone-spassword/pull/57/commits/7eeef74282622cfb58b1b66d7a24eca8a7390359

File: docs/second_factor_auth.md (hunk 3 above)
Comment (AlvaroVega) [reply docs/second_factor_auth.md]: fixed in https://github.com/telefonicaid/fiware-keystone-spassword/pull/57/commits/20ef18a09749b0702a2a88d40475194fb490d370

**Pull request merged**